batch.py

```diff
@@ -0,0 +1,389 @@
+    mt = hl.read_matrix_table(mt_path)
+
+    # subset to relevant samples (samples we have scRNA-seq data for)
+    mt = mt.filter_cols(hl.set(samples).contains(mt.s))
+
+    # densify (can this be done at the end?)
+    mt = hl.experimental.densify(mt)
+
+    # filter out low quality variants and consider biallelic variants only (no multi-allelic, no ref-only)
+    mt = mt.filter_rows(  # check these filters!
+        (hl.len(hl.or_else(mt.filters, hl.empty_set(hl.tstr))) == 0)  # QC
+        & (hl.len(mt.alleles) == 2)  # remove hom-ref
+        & (mt.n_unsplit_alleles == 2)  # biallelic
+        & (hl.is_snp(mt.alleles[0], mt.alleles[1]))  # SNVs
+    )
+
+    # filter rare variants only (MAF < 5%)
+    mt = hl.variant_qc(mt)
+    mt = mt.filter_rows(
+        (mt.variant_qc.AF[1] < 0.05) & (mt.variant_qc.AF[1] > 0)
+        | (mt.variant_qc.AF[1] > 0.95) & (mt.variant_qc.AF[1] < 1)
+    )
+    mt = mt.checkpoint(output_mt_path, overwrite=True)
+    logging.info(
+        f"Number of rare (freq<5%) and QC'd biallelic variants: {mt.count()[0]}"
+    )
+
+
+# endregion SUBSET_VARIANTS
+
+
+# region GET_GENE_SPECIFIC_VARIANTS
+
+
+def get_promoter_variants(
+    mt_path: str,  # checkpoint from function above
+    ht_path: str,
+    gene_details: dict[str, str],  # ouput of make_gene_loc_dict
+    window_size: int,
+    plink_output_prefix: str,  # 'tob_wgs_rv/pseudobulk_rv_association/plink_files/'
+):
+    """Subset hail matrix table
+
+    Input:
+    mt_path: path to already subsetted hail matrix table
+    gene_file: path to a tsv file with information on
+    the chrom, start and end location of genes (rows)
+
+    Output:
+    For retained variants, that are: 1) in promoter regions and
+    2) within 50kb up or down-stream of the gene body (or in the gene body itself)
+    (on top of all filters done above)
+
+    returns:
+    - Path to plink prefix for genotype files (plink format: .bed, .bim, .fam)
+    - Path to hail table with variant (rows) stats for downstream analyses
+    """
+    # read hail matrix table object (pre-filtered)
+    init_batch()
+    mt = hl.read_matrix_table(mt_path)
+
+    gene_name = gene_details["gene_name"]
+
+    # get relevant chromosome
+    chrom = gene_details["chrom"]
+
+    # subset to window
+    # get gene body position (start and end) and build interval
+    left_boundary = max(1, int(gene_details["gene-start"]) - window_size)
+    right_boundary = min(
+        int(gene_details["gene-end"]) + window_size,
+        hl.get_reference("GRCh38").lengths[f"chr{chrom}"],
+    )
+    # get gene-specific genomic interval
+    gene_interval = f"{chrom}:{left_boundary}-{right_boundary}"
+    logging.info(f"Interval considered: {gene_interval}")  # 'chr22:23219960-23348287'
+
+    # include variants up to {window size} up- and downstream
+    mt = hl.filter_intervals(
+        mt, [hl.parse_locus_interval(gene_interval, reference_genome="GRCh38")]
+    )
+    mt_path = output_path("in_window.mt", "tmp")
```

Comment:
```suggestion
    mt_path = output_path(f"{gene_name}_in_window.mt", "tmp")
```

Interested in knowing if this is a useful checkpoint - the only filtering done up to this point is on a table key (Chr-Pos), so I would expect hail to execute that without much computational overhead. It's good to remember that checkpointing isn't 'free', but I don't have a good idea of how much (if at all) it will speed up the annotation step